BVAL-559 Adding NoProviderFoundException; BVAL-264 Improving ConstraintViolationException message

gunnarmorling opened the pull request:
https://hibernate.atlassian.net/projects/BVAL/issues/BVAL-559
https://hibernate.atlassian.net/projects/BVAL/issues/BVAL-264

Spec change coming up, too. Let's leave both open for a few days to get feedback from the EG. I'll also ask people to take a look on the mailing list once I've added the spec change.

chkal commented:
Looks good. +1

gsmet commented:
Looks good to me.

gunnarmorling commented:
Pushed another commit for BVAL-264.

otaviojava commented on src/main/java/javax/validation/ConstraintViolationException.java:
```diff
@@ -32,7 +34,7 @@ public ConstraintViolationException(String message,
 		}
 		else {
-			this.constraintViolations = new HashSet<ConstraintViolation<?>>( constraintViolations );
+			this.constraintViolations = new HashSet<>( constraintViolations );
```
:question: why no just make `constraintViolations` immutable?
Eg: ` java.util.Collections.unmodifiableSet`

chkal replied:
Good point.

gunnarmorling replied:
I had been thinking about this, too. I refrained from doing it as it's behavioral change that's not strictly needed. But then the spec makes no statement on the mutability of this set, so we should be fine to change it.

chkal replied:
I agree that it is a behavioral change and therefore perhaps doesn't belong into this PR. 

However, I also think that nobody is expecting this `Set` to be mutable, so perhaps we should make it immutable AND adjust the javadocs in a separate PR?

gunnarmorling replied:
Thinking more about it, I don't see any real advantage of making it immutable. It's not that we'd share the set of violations between multiple invocations or threads, it's specific to one invocation of the validation engine. And maybe some user wants to amend the exception when passing it up. It seems far fetched, but I don't think we have strong reason to prevent it.

chkal replied:
I guess it is a matter of taste. For me it feels odd that you can do something like:

```java
constraintViolationException.getConstraintViolations().add(null);
```

However, I'm not strong on this one. I'm fine with leaving it as it is.

gunnarmorling replied:
Agreed it's odd, and I also cannot see a case where I'd do it. But then I don't see a good reason to rule it out. The one example I can see is that you catch the exception, add another constraint violation with some additional info and re-throw it. It's not a usage I'd encourage, but think it's alright to allow it.

otaviojava commented on src/main/java/javax/validation/Validation.java:
```diff
@@ -273,7 +275,7 @@ public ValidationProviderResolver getDefaultValidationProviderResolver() {
 			if ( validationProviders.size() == 0 ) {
```
:warning: I believe `if (validationProviders.isEmpty)` is more readeable

gunnarmorling replied:
Good idea!

otaviojava commented on src/main/java/javax/validation/Validation.java:
```diff
@@ -273,7 +275,7 @@ public ValidationProviderResolver getDefaultValidationProviderResolver() {
 			if ( validationProviders.size() == 0 ) {
 				String msg = "Unable to create a Configuration, because no Bean Validation provider could be found." +
```
Maybe put it as a constant here.

emmanuelbernard replied:
What difference does it make in the end? Constants are copied inline by the compiler.

gunnarmorling replied:
Agreed, it won't change much. If anything, it should be done for all the messages used in this class. @otaviojava, if like, send in a PR for that :)

gunnarmorling commented:
Rebased and applied. Thanks for reviewing everyone. I've addressed the remark about `isEmpty()`, about the others I think we are better off with the original approach.
